mozetl/topline/topline_dashboard.py

```diff
@@ -0,0 +1,197 @@
+    # schema. Columns missing in the topline_summary from the
+    # historical dataset are aggregates, so they are are filled with
+    # zeroes.
+    def default_column(name):
+        return F.lit(0).alias(name)
+
+    column_expr = [name if name in topline_columns else default_column(name)
+                   for name in historical_schema.names]
+    formatted_df = filtered_df.select(*column_expr)
+
+    return formatted_df
+
+
+def write_dashboard_data(df, bucket, prefix, mode):
+    """ Write the dashboard data to a s3 location. """
+
+    # create a temporary directory to dump files into
+    path = tempfile.mkdtemp()
+
+    # write dataframe to working directory
+    df.repartition(1).write.csv(path, header=True, mode="overwrite")
+
+    # find the file that was created
+    filepath = None
+    for name in os.listdir(path):
+        if name.endswith(".csv"):
+            filepath = os.path.join(path, name)
+            break
+
+    # name of the output key
+    key = "{}/v4-{}.csv".format(prefix, mode)
+
+    # create the s3 resource for this transaction
+    s3 = boto3.client('s3', region_name='us-west-2')
+
+    # write the contents of the file to right location
+    with open(filepath, 'rb') as data:
+        s3.put_object(Bucket=bucket,
+                      Key=key,
+                      Body=data,
+                      ACL='bucket-owner-full-control')
+
+    logger.info('Sucessfully wrote {} to {}'.format(key, bucket))
+
+    # clean up the temporary directory
+    shutil.rmtree(path)
+
+
+@click.command()
+@click.argument('report_start')
+@click.argument('mode', type=click.Choice(['weekly', 'monthly']))
+@click.argument('bucket')
+@click.argument('prefix')
+@click.option('--input_bucket',
+              default='telemetry-parquet',
+              help='Bucket of the ToplineSummary dataset')
+@click.option('--input_prefix',
+              default='topline_summary/v1',
+              help='Prefix of the ToplineSummary dataset')
+def main(report_start, mode, bucket, prefix, input_bucket, input_prefix):
+    spark = (SparkSession
+             .builder
+             .appName("topline_dashboard")
+             .getOrCreate())
+
+    logger.info('Generating {} topline_dashboard for {}'
+                .format(mode, report_start))
+
+    # the inclusion of mode doesn't matter, but we need report_start
+    input_path = format_spark_path(
+        input_bucket,
+        "{}/mode={}".format(input_prefix, mode)
+    )
+    logger.info('Reading input data from {}'.format(input_path))
+
+    # Note: The schema is applied to cast report_start to a string. The
+    # DataFrameReader interface has a `.schema()` that mysteriously fails
+    # in this context, so the schema is applied after reading the dataframe
+    # into memory.
+    topline_summary = (
+        spark.read
+        .parquet(input_path)
+        .rdd.toDF(topline_schema)
+        .where(F.col('report_start') == report_start)
```

Comment:
I'm just going to remove report_start, since it doesn't look like its going to see much use. I think this was left-over from my initial sketch that was going to more closely follow the run.sh script.